src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc

```diff
@@ -608,43 +610,54 @@ bool ShouldUseAres(const char* resolver_env) {
          gpr_stricmp(resolver_env, "ares") == 0;
 }
 
-bool g_use_ares_dns_resolver;
+bool UseAresDnsResolver() {
+  static const bool result = []() {
+    UniquePtr<char> resolver = GPR_GLOBAL_CONFIG_GET(grpc_dns_resolver);
+    bool result = ShouldUseAres(resolver.get());
+    if (result) gpr_log(GPR_DEBUG, "Using ares dns resolver");
+    return result;
+  }();
+  return result;
+}
 
 }  // namespace
 
+void RegisterAresDnsResolver(CoreConfiguration::Builder* builder) {
+  if (UseAresDnsResolver()) {
+    builder->resolver_registry()->RegisterResolverFactory(
+        absl::make_unique<AresClientChannelDNSResolverFactory>());
+  }
+}
+
 }  // namespace grpc_core
 
 void grpc_resolver_dns_ares_init() {
-  grpc_core::UniquePtr<char> resolver =
-      GPR_GLOBAL_CONFIG_GET(grpc_dns_resolver);
-  if (grpc_core::ShouldUseAres(resolver.get())) {
-    grpc_core::g_use_ares_dns_resolver = true;
-    gpr_log(GPR_DEBUG, "Using ares dns resolver");
+  if (grpc_core::UseAresDnsResolver()) {
     address_sorting_init();
     grpc_error_handle error = grpc_ares_init();
     if (error != GRPC_ERROR_NONE) {
       GRPC_LOG_IF_ERROR("grpc_ares_init() failed", error);
       return;
     }
     grpc_core::SetDNSResolver(grpc_core::AresDNSResolver::GetOrCreate());
-    grpc_core::ResolverRegistry::Builder::RegisterResolverFactory(
-        absl::make_unique<grpc_core::AresClientChannelDNSResolverFactory>());
-  } else {
-    grpc_core::g_use_ares_dns_resolver = false;
   }
 }
 
 void grpc_resolver_dns_ares_shutdown() {
-  if (grpc_core::g_use_ares_dns_resolver) {
+  if (grpc_core::UseAresDnsResolver()) {
     address_sorting_shutdown();
     grpc_ares_cleanup();
   }
 }
 
 #else /* GRPC_ARES == 1 */
 
-void grpc_resolver_dns_ares_init(void) {}
+namespace grpc_core {
+void RegisterAresDnsResolver(CoreConfiguration::Builder*) {}
+}  // namespace grpc_core
+
+void grpc_resolver_dns_ares_init() {}
```

Comment:
Are these two functions still needed?

Reply:
Yeah... when we're using ares we need to init address sorting and something else... which can't yet go into core configuration (next steps...)